src/plugins/embeddable/server/plugin.ts

```diff
@@ -0,0 +1,186 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { CoreSetup, CoreStart, Plugin, SavedObjectReference } from 'kibana/server';
+import { identity } from 'lodash';
+import {
+  EmbeddableFactoryRegistry,
+  EnhancementsRegistry,
+  EnhancementRegistryDefinition,
+  EnhancementRegistryItem,
+  EmbeddableRegistryDefinition,
+} from './types';
+import {
+  extractBaseEmbeddableInput,
+  injectBaseEmbeddableInput,
+  telemetryBaseEmbeddableInput,
+} from '../common/lib/migrate_base_input';
+import { SerializableState } from '../../kibana_utils/common';
+import { EmbeddableInput } from '../common/types';
+
+export interface EmbeddableSetup {
+  registerEmbeddableFactory: (factory: EmbeddableRegistryDefinition) => void;
+  registerEnhancement: (enhancement: EnhancementRegistryDefinition) => void;
+}
+
+export class EmbeddableServerPlugin implements Plugin<object, object> {
+  private readonly embeddableFactories: EmbeddableFactoryRegistry = new Map();
+  private readonly enhancements: EnhancementsRegistry = new Map();
+
+  public setup(core: CoreSetup) {
+    return {
+      registerEmbeddableFactory: this.registerEmbeddableFactory,
+      registerEnhancement: this.registerEnhancement,
+    };
+  }
+
+  public start(core: CoreStart) {
+    return {
+      telemetry: this.telemetry,
+      extract: this.extract,
+      inject: this.inject,
+    };
+  }
+
+  public stop() {}
+
+  private telemetry = (state: EmbeddableInput, telemetryData: Record<string, any> = {}) => {
+    const enhancements: Record<string, any> = state.enhancements || {};
+    const factory = this.getEmbeddableFactory(state.id);
+
+    telemetryBaseEmbeddableInput(state, telemetryData);
+    if (factory) {
+      factory.telemetry(state, telemetryData);
```

Comment:
This implementation seems strange to me - `telemetry` returns an updated `telemetryData` object, but we are simply discarding and probably assume the passed in object will be mutated.

I really don't like relying on object reference here and mutation the object passed into the function. This is not how reduce is typically implemented - can we make `telemetry` return a new object with added information instead of mutating the existing one? This can lead to hard to catch bugs.